### PR TITLE
Add VPN IPsec Phase 1 interface

### PR DESCRIPTION
**Pull request by alagoutte**

### CODE REVIEW

File: PowerFGT/Private/Confirm.ps1 [hunk 1]
```diff
@@ -336,3 +336,28 @@ Function Confirm-FGTInterface {
...
 }
+
+Function Confirm-FGTVpnIpsecPhase1Interface {
+
+    Param (
+        [Parameter (Mandatory = $true)]
+        [object]$argument
+    )
+
+    #Check if it looks like a VPN IPsec Phase 1 Interface  element
```
Comment (alagoutte) [inline PowerFGT/Private/Confirm.ps1:347]: Extra space

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 2]
```diff
@@ -3,6 +3,227 @@
...
+        .EXAMPLE
+        Add-FGTVpnIpsecPhase1Interface -name PowerFGT_VPN -type dynamic -interface port2 -proposal aes256-sha256, aes256-sha512 -dhgrp 14,15 -psksecret MySecret
+
+        Create a dynamic VPN IPsec Phase 1 Interface named PowerFGT_VPN with interface port2, multiple proposal aes256-sha256, aes256-sha512 and DH Group 14 & 15
+
+        .EXAMPLE
+        $data = @{ "fragmentation" = "disable" ; "npu-offload" = "disable" }
+        PS C> Add-FGTVpnIpsecPhase1Interface -name PowerFGT_VPN -type static -interface port2 -psksecret MySecret -remotegw 192.0.2.1 -data $data
+
+        reate a dynamic VPN IPsec Phase 1 Interface named PowerFGT_VPN  with fragmentation and npu-offload using -data parameter
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:30]: Extra space

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 3]
```diff
@@ -3,6 +3,227 @@
...
+                $_interface | Add-member -name "exchange-interface-ip" -membertype NoteProperty -Value "enable"
+            }
+            else {
+                $_interface | Add-member -name "exchange-interface-ip" -membertype NoteProperty -Value "disable"
+            }
+        }
+
+        if ( $PsBoundParameters.ContainsKey('networkid') ) {
+            if ($ikeversion -eq "2") {
+
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:182]: Remove blank line

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 4]
```diff
@@ -112,3 +333,288 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+        .SYNOPSIS
+        Configure a Vpn IPsec Phase 1 Interface
+
+        .DESCRIPTION
+        Configure an Vpn IPsec Phase 1 Interface (Version, type, interface, proposal, psksecret... )
+
+        .EXAMPLE
+        Get-FGTVpnIpsecPhase1Interface PowerFGT_VPN | Set-FGTVpnIpsecPhase1Interface -interface port2 -psksecret MySecret -remotegw 192.0.2.1
+
+        Create a static VPN IPsec Phase 1 Interface named PowerFGT_VPN with interface port2 with Remote Gateway 192.0.2.1
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:349]: Fix example description (x3)

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 5]
```diff
@@ -112,3 +333,288 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+
+        .EXAMPLE
+        Add-FGTVpnIpsecPhase1Interface -name PowerFGT_VPN -type dynamic -interface port2 -proposal aes256-sha256, aes256-sha512 -dhgrp 14,15 -psksecret MySecret
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:352]: Fix example command

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 6]
```diff
@@ -112,3 +333,288 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+        if ( $PsBoundParameters.ContainsKey('vdom') ) {
+            $invokeParams.add( 'vdom', $vdom )
+        }
+
+        $uri = "api/v2/cmdb/vpn.ipsec/phase1-interface/$($vpn.name)"
+
+        $_interface = new-Object -TypeName PSObject
+
+        if ( $PsBoundParameters.ContainsKey('name') ) {
+
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:435]: Remove extra blank line

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 7]
```diff
@@ -112,3 +333,288 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+            $_interface | add-member -name "name" -membertype NoteProperty -Value $name
+        }
+        if ( $PsBoundParameters.ContainsKey('interface') ) {
+
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:439]: Same here about extra blank line

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 8]
```diff
@@ -112,3 +333,288 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+            $_interface | add-member -name "interface" -membertype NoteProperty -Value $interface
+        }
+
+        if ( $PsBoundParameters.ContainsKey('psksecret') ) {
+            $_interface | add-member -name "psksecret" -membertype NoteProperty -Value $psksecret
+        }
+
+        if ( $PsBoundParameters.ContainsKey('ikeversion') ) {
+            $_interface | add-member -name "ike-version" -membertype NoteProperty -Value $ikeversion
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:448]: Really working to change Ike-version ?

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,985 @@
...
+        Add-FGTVpnIpsecPhase1Interface -name $pester_vpn2 -type dynamic -interface port2 -proposal aes256-sha256 -psksecret MySecret
+    }
+
+    It "Get VPN Ipsec Phase 1 interface does not throw an error" {
+        {
+            Get-FGTVpnIpsecPhase1Interface
+        } | Should -Not -Throw
+    }
+
+    It "Get ALL VPN Ipsec Phase 1 interfaces" {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:27]: Inteface(s)

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,985 @@
...
+        $interface = Get-FGTVpnIpsecPhase1Interface
+        $interface.count | Should -Not -Be $NULL
+    }
+
+    It "Get ALL VPN Ipsec Phase 1 interface with -skip" {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:32]: Interface(s)

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,985 @@
...
+            if ($_.param.type -eq "static") {
+                $vpn.'remote-gw' | Should -Be "192.0.2.1"
+            }
+            else {
+                $vpn.'remote-gw' | Should -Be "0.0.0.0"
+            }
+            $vpn.dhgrp | Should -Be "1 2"
+        }
+
+        It "Set VPN Ipsec Phase 1 Interface with net-device enabled" -skip:($fgt_version -lt "6.2.0" -and $_.param.type -eq "static") {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:631]: Set disable after

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,985 @@
...
+            if ($_.param.type -eq "static") {
+                $vpn.'remote-gw' | Should -Be "192.0.2.1"
+            }
+            else {
+                $vpn.'remote-gw' | Should -Be "0.0.0.0"
+            }
+            $vpn.'net-device' | Should -Be "enable"
+        }
+
+        It "Set VPN Ipsec Phase 1 Interface with add-route disabled" {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:648]: Set enable after

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 13]
```diff
@@ -0,0 +1,985 @@
...
+            }
+            else {
+                $vpn.'remote-gw' | Should -Be "0.0.0.0"
+            }
+            $vpn.'dpd' | Should -Be "on-demand"
+            $vpn.'dpd-retrycount' | Should -Be 1
+            $vpn.'dpd-retryinterval' | Should -Be 10
+        }
+
+        It "Set VPN Ipsec Phase 1 Interface with idle timeout enabled" {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:905]: Set disable after

**Review by alagoutte — COMMENTED** [review]
Need some change before merge and also update readme

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 14]
```diff
@@ -3,6 +3,226 @@
...
 #
+
+function Add-FGTVpnIpsecPhase1Interface {
+
+    <#
+        .SYNOPSIS
+        Add a Vpn IPsec Phase 1 Interface
+
+        .DESCRIPTION
+        Add an Vpn IPsec Phase 1 Interface (Version, type, interface, proposal, psksecret... )
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:14]: an -> a

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 15]
```diff
@@ -0,0 +1,1050 @@
...
+        }
+
+        AfterAll {
+            Get-FGTVpnIpsecPhase1Interface -name $pester_vpn1 | Remove-FGTVpnIpsecPhase1Interface -Confirm:$false
+        }
+    }
+
+}
+
+Describe "Add VPN Ipsec Phase 1 Interface" -ForEach $type {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:1031]: Add _> Remove

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 16]
```diff
@@ -0,0 +1,1050 @@
...
+        $interface = Get-FGTVpnIpsecPhase1Interface
+        $interface.count | Should -Not -Be $NULL
+    }
+
+    It "Get ALL VPN Ipsec Phase 1 interface(s) with -skip" {
+        $interface = Get-FGTVpnIpsecPhase1Interface -skip
+        $interface.count | Should -Not -Be $NULL
+    }
+
+    It "Get VPN Ipsec Phase 1 interface ($pester_int1)" {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:37]: $pester_int1 -> $pester_vpn1

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 17]
```diff
@@ -0,0 +1,1050 @@
...
+Describe "Add VPN Ipsec Phase 1 Interface" -ForEach $type {
+
+    Context "Interface $($_.type)" {
+
+        BeforeEach {
+            $p = $_.param
+            Add-FGTVpnIpsecPhase1Interface -name $pester_vpn1 -interface $pester_port1 -psksecret MySecret @p
+        }
+
+        It "Remove System Interface by pipeline" {
```
Comment (alagoutte) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:1040]: system -> VPN Ipsec Phase 1

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 (hunk 8 above)
Comment (alagoutte) [reply PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:445]: Yes ! (And there is now test...)

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 18]
```diff
@@ -112,3 +332,286 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+        $invokeParams = @{ }
+        if ( $PsBoundParameters.ContainsKey('vdom') ) {
+            $invokeParams.add( 'vdom', $vdom )
+        }
+
+        $uri = "api/v2/cmdb/vpn.ipsec/phase1-interface/$($vpn.name)"
+
+        $_interface = new-Object -TypeName PSObject
+
+        if ( $PsBoundParameters.ContainsKey('name') ) {
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:433]: to test for change name...

File: PowerFGT/Private/Confirm.ps1 [hunk 19]
```diff
@@ -336,3 +336,28 @@ Function Confirm-FGTInterface {
...
+
+Function Confirm-FGTVpnIpsecPhase1Interface {
+
+    Param (
+        [Parameter (Mandatory = $true)]
+        [object]$argument
+    )
+
+    #Check if it looks like a VPN IPsec Phase 1 Interface element
+
```
Comment (CedricMoreau) [inline PowerFGT/Private/Confirm.ps1:348]: remove blank line

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 20]
```diff
@@ -3,6 +3,226 @@
...
+        .DESCRIPTION
+        Add an Vpn IPsec Phase 1 Interface (Version, type, interface, proposal, psksecret... )
+
+        .EXAMPLE
+        Add-FGTVpnIpsecPhase1Interface -name PowerFGT_VPN -type static -interface port2 -psksecret MySecret -remotegw 192.0.2.1
+
+        Create a static VPN IPsec Phase 1 Interface named PowerFGT_VPN with interface port2 with Remote Gateway 192.0.2.1
+
+        .EXAMPLE
+        Add-FGTVpnIpsecPhase1Interface -name PowerFGT_VPN -type dynamic -interface port2 -proposal aes256-sha256, aes256-sha512 -dhgrp 14,15 -psksecret MySecret
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:22]: -proposal aes256-sha256, aes256-sha512 --> remove white space after comma

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 21]
```diff
@@ -3,6 +3,226 @@
...
+
+        Create a dynamic VPN IPsec Phase 1 Interface named PowerFGT_VPN with interface port2, multiple proposal aes256-sha256, aes256-sha512 and DH Group 14 & 15
+
+        .EXAMPLE
+        $data = @{ "fragmentation" = "disable" ; "npu-offload" = "disable" }
+        PS C> Add-FGTVpnIpsecPhase1Interface -name PowerFGT_VPN -type static -interface port2 -psksecret MySecret -remotegw 192.0.2.1 -data $data
+
+        Create a dynamic VPN IPsec Phase 1 Interface named PowerFGT_VPN with fragmentation and npu-offload using -data parameter
+
+    #>
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:32]: Add blank line?

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 22]
```diff
@@ -3,6 +3,226 @@
...
+        [String[]]$vdom,
+        [Parameter(Mandatory = $false)]
+        [psobject]$connection = $DefaultFGTConnection
+    )
+
+    Begin {
+    }
+
+    Process {
+
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:91]: remove blank line ?

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 23]
```diff
@@ -3,6 +3,226 @@
...
+        $invokeParams = @{ }
+        if ( $PsBoundParameters.ContainsKey('vdom') ) {
+            $invokeParams.add( 'vdom', $vdom )
+        }
+
+        $uri = "api/v2/cmdb/vpn.ipsec/phase1-interface"
+        $_interface = new-Object -TypeName PSObject
+
+        $_interface | add-member -name "name" -membertype NoteProperty -Value $name
+
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:101]: blank line ?

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 24]
```diff
@@ -3,6 +3,226 @@
...
+        if ( $PsBoundParameters.ContainsKey('dhgrp') ) {
+            $_interface | add-member -name "dhgrp" -membertype NoteProperty -Value ($dhgrp -join " ")
+        }
+
+        if ( $PsBoundParameters.ContainsKey('netdevice') ) {
+            if ($netdevice) {
+                $_interface | Add-member -name "net-device" -membertype NoteProperty -Value "enable"
+            }
+            else {
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:134]: The else should be on the primary If ? If netdevice is a parameter, you set its value to enable, else it's not a parameter so you set it to disable ?

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 25]
```diff
@@ -112,3 +332,286 @@ function Get-FGTVpnIpsecPhase1Interface {
...
 }
+
+function Set-FGTVpnIpsecPhase1Interface {
+
+    <#
+        .SYNOPSIS
+        Configure a Vpn IPsec Phase 1 Interface
+
+        .DESCRIPTION
+        Configure an Vpn IPsec Phase 1 Interface (Version, type, interface, proposal, psksecret... )
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:343]: an --> a

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 26]
```diff
@@ -112,3 +332,286 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+
+        .EXAMPLE
+        Get-FGTVpnIpsecPhase1Interface PowerFGT_VPN | Set-FGTVpnIpsecPhase1Interface -psksecret MySecret
+
+        Change psksecret of VPN IPsec Phase1 Interface PowerFGT_VPN
+
+        .EXAMPLE
+        Get-FGTVpnIpsecPhase1Interface PowerFGT_VPN | Set-FGTVpnIpsecPhase1Interface -proposal aes256-sha256, aes256-sha512 -dhgrp 14,1
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:351]: white space ?

File: PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1 [hunk 27]
```diff
@@ -112,3 +332,286 @@ function Get-FGTVpnIpsecPhase1Interface {
...
+        [Parameter(Mandatory = $false)]
+        [psobject]$connection = $DefaultFGTConnection
+    )
+
+    Begin {
+    }
+
+    Process {
+
+
```
Comment (CedricMoreau) [inline PowerFGT/Public/cmdb/vpn/ipsec/phase1-interface.ps1:423]: blank line ?

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 28]
```diff
@@ -0,0 +1,1050 @@
...
+}
+
+$type = @(
+    @{ "type" = "IKEv1 Static"; "param" = @{ "ikeversion" = 1; "type" = "static"; "remotegw" = "192.0.2.1" } }
+    @{ "type" = "IKEv1 Dynamic"; "param" = @{ "ikeversion" = 1; "type" = "dynamic" } }
+    @{ "type" = "IKEv2 Static"; "param" = @{ "ikeversion" = 2; "type" = "static" ; "remotegw" = "192.0.2.1" } }
+    @{ "type" = "IKEv2 Dynamic"; "param" = @{ "ikeversion" = 2; "type" = "dynamic" } }
+)
+
+
```
Comment (CedricMoreau) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:71]: blank line ?

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 29]
```diff
@@ -0,0 +1,1050 @@
...
+            }
+            else {
+                $vpn.'remote-gw' | Should -Be "0.0.0.0"
+            }
+            $vpn.proposal | Should -Be "des-sha1"
+        }
+
+        It "Add VPN Ipsec Phase 1 Interface with 2 proposals (des-md5, des-sha1)" {
+            $p = $_.param
+            Add-FGTVpnIpsecPhase1Interface -name $pester_vpn1 -interface $pester_port1 -psksecret MySecret @p -proposal des-md5, des-sha1
```
Comment (CedricMoreau) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:153]: white space not breaking proposal ?

File: Tests/integration/VpnIPsecPhase1Interface.Tests.ps1 [hunk 30]
```diff
@@ -0,0 +1,1050 @@
...
+            }
+            else {
+                $vpn.'remote-gw' | Should -Be "0.0.0.0"
+            }
+            $vpn.dhgrp | Should -Be "2"
+        }
+
+        It "Add VPN Ipsec Phase 1 Interface with 2 DH Group (1, 2)" {
+            $p = $_.param
+            Add-FGTVpnIpsecPhase1Interface -name $pester_vpn1 -interface $pester_port1 -psksecret MySecret @p -dhgrp 1, 2
```
Comment (CedricMoreau) [inline Tests/integration/VpnIPsecPhase1Interface.Tests.ps1:206]: white space not breaking dh groupe ?

**Pull request merged**